subsys/testsuite/ztest/src/ztest_new.c

```diff
@@ -598,26 +595,28 @@ void end_report(void)
 K_APPMEM_PARTITION_DEFINE(ztest_mem_partition);
 #endif
 
-int ztest_run_test_suites(const void *state)
+int z_impl_ztest_run_test_suites(const void *state)
 {
 	struct ztest_suite_node *ptr;
 	int count = 0;
 
 #ifdef CONFIG_ZTEST_SHUFFLE
-	struct ztest_suite_node tmp;
-
-	z_ztest_shuffle(_ztest_suite_node_list_start,
-			_ztest_suite_node_list_end - _ztest_suite_node_list_start, &tmp,
-			sizeof(struct ztest_suite_node));
-#endif
+	struct ztest_suite_node *suites_to_run[ZTEST_SUITE_COUNT];
 
+	memset(suites_to_run, 0, ZTEST_SUITE_COUNT * sizeof(struct ztest_suite_node *));
+	z_ztest_shuffle((void **)suites_to_run, (intptr_t)_ztest_suite_node_list_start,
+			ZTEST_SUITE_COUNT, sizeof(struct ztest_suite_node));
+	for (size_t i = 0; i < ZTEST_SUITE_COUNT; ++i) {
+		ptr = suites_to_run[i];
+#else
 	for (ptr = _ztest_suite_node_list_start; ptr < _ztest_suite_node_list_end; ++ptr) {
-		struct ztest_suite_stats *stats = &ptr->stats;
+#endif
+		struct ztest_suite_stats *stats = ptr->stats;
```

Comment:
NIT, I'm not a fan of the for loop being split by the conditional `ZTEST_SHUFFLE`. Could we make `suites_to_run` work outside of the ifdefs?